test_load_controller tests always return true

> @christophfroehlich it's probable that they are returning successful even without parameters. That's the only explanation. I could try fixing them tomorrow.

but it throws the exception, and CM::add_controller_impl should return nullptr..

Furthermore,  I'm not sure why load_JTC succeeds. it should throw an error without paramaters, but I only see 
```
1: [INFO] [1724185010.487326696] [test_controller_manager]: Subscribing to '/robot_description' topic for robot description.
1: [INFO] [1724185010.487412910] [test_controller_manager]: Loading controller 'test_joint_trajectory_controller'
1: [ERROR] [1724185010.496710000] [test_joint_trajectory_controller]: Unable to start transition 5 from current state shuttingdown: Could not publish transition: publisher's context is invalid, at ./src/rcl/publisher.c:423, at ./src/rcl_lifecycle.c:368
1: [WARN] [1724185010.496722342] [rclcpp_lifecycle]: Shutdown error in destruction of LifecycleNode: final state(unconfigured)
1: [       OK ] TestLoadJointStateController.load_controller (40 ms)
```

This is not really linked to this PR here. I suggest applying the same changes to admittance_ctrl and range_sensor_broadcaster (the only controllers using a yaml file which are not already addressed) and create a follow-up issue for those false-positive tests.

_Originally posted by @christophfroehlich in https://github.com/ros-controls/ros2_controllers/issues/1256#issuecomment-2299717499_

--- a/motion_primitives_controllers/CMakeLists.txt
+++ b/motion_primitives_controllers/CMakeLists.txt
@@ -84,6 +84,8 @@ if(BUILD_TESTING)
   find_package(hardware_interface REQUIRED)
   find_package(ros2_control_test_assets REQUIRED)
 
+  add_definitions(-DTEST_FILES_DIRECTORY="${CMAKE_CURRENT_SOURCE_DIR}/test")
+
   ament_add_gmock(
     test_load_motion_primitives_forward_controller
     test/test_load_motion_primitives_forward_controller.cpp
@@ -91,6 +93,8 @@ if(BUILD_TESTING)
   target_link_libraries(test_load_motion_primitives_forward_controller
     ${PROJECT_NAME}
     controller_manager::controller_manager
+  )
+  target_link_libraries(test_load_motion_primitives_forward_controller
     ros2_control_test_assets::ros2_control_test_assets
   )
 
@@ -103,6 +107,8 @@ if(BUILD_TESTING)
     ${PROJECT_NAME}
     controller_interface::controller_interface
     hardware_interface::hardware_interface
+  )
+  target_link_libraries(test_motion_primitives_forward_controller
     ros2_control_test_assets::ros2_control_test_assets
   )
 endif()

--- a/motion_primitives_controllers/test/test_load_motion_primitives_forward_controller.cpp
+++ b/motion_primitives_controllers/test/test_load_motion_primitives_forward_controller.cpp
@@ -32,9 +32,13 @@ TEST(TestLoadMotionPrimitivesForwardController, load_controller)
   controller_manager::ControllerManager cm(
     executor, ros2_control_test_assets::minimal_robot_urdf, true, "test_controller_manager");
 
-  ASSERT_NO_THROW(cm.load_controller(
-    "test_motion_primitives_forward_controller",
-    "motion_primitives_controllers/MotionPrimitivesForwardController"));
+  const std::string test_file_path =
+    std::string(TEST_FILES_DIRECTORY) + "/motion_primitives_forward_controller_params.yaml";
+  cm.set_parameter({"test_motion_primitives_forward_controller.params_file", test_file_path});
+  cm.set_parameter(
+    {"test_motion_primitives_forward_controller.type",
+     "motion_primitives_controllers/MotionPrimitivesForwardController"});
+  ASSERT_NE(cm.load_controller("test_motion_primitives_forward_controller"), nullptr);
 
   rclcpp::shutdown();
 }

--- a/pid_controller/CMakeLists.txt
+++ b/pid_controller/CMakeLists.txt
@@ -67,6 +67,8 @@ if(BUILD_TESTING)
   find_package(controller_manager REQUIRED)
   find_package(ros2_control_test_assets REQUIRED)
 
+  add_definitions(-DTEST_FILES_DIRECTORY="${CMAKE_CURRENT_SOURCE_DIR}/test")
+
   add_rostest_with_parameters_gmock(
     test_pid_controller
     test/test_pid_controller.cpp

--- a/pid_controller/test/test_load_pid_controller.cpp
+++ b/pid_controller/test/test_load_pid_controller.cpp
@@ -35,7 +35,11 @@ TEST(TestLoadPidController, load_controller)
   controller_manager::ControllerManager cm(
     executor, ros2_control_test_assets::minimal_robot_urdf, true, "test_controller_manager");
 
-  ASSERT_NO_THROW(cm.load_controller("test_pid_controller", "pid_controller/PidController"));
+  const std::string test_file_path =
+    std::string(TEST_FILES_DIRECTORY) + "/pid_controller_params.yaml";
+  cm.set_parameter({"test_pid_controller.params_file", test_file_path});
+  cm.set_parameter({"test_pid_controller.type", "pid_controller/PidController"});
+  ASSERT_NE(cm.load_controller("test_pid_controller"), nullptr);
 
   rclcpp::shutdown();
 }

